src/dump.rs

```diff
@@ -0,0 +1,281 @@
+//
+// This implies that there's a full clone of the dump held on for those leaf-specific views, which
+// is unfortunate! It's also not yet really clear how to assemble those more complex leaves for
+// writer purposes.
+#[derive(Clone)]
+pub struct CpuIdDump {
+    leaves: HashMap<u32, LeafOrSubleaves>,
+}
+
+impl CpuIdDump {
+    // TODO: probably should just take vendor in the initial constructor here
+    // (that also lets this pick the right leaf/subleaf fallback behavior from the get-go)
+    pub fn new() -> Self {
+        Self {
+            leaves: HashMap::new(),
+        }
+    }
+}
+
+pub struct CpuIdDumpIter {
+    // It's straightforward enough to use `hash_map::Drain` to walk the top-level map but it's more
+    // annoying for inner collections of subleaves because `Drain` holds a borrow of the
+    // to-be-drained map. Here, that'd mean the struct is self-referential with `current_subleaf`
+    // borrowing `dump`. So, just be naive the whole way through (much to the dismay of `impl
+    // Iterator` below..)
+    dump: CpuIdDump,
+    leaf: u32,
+    current_subleaf: Option<HashMap<u32, CpuIdResult>>,
+}
+
+impl IntoIterator for CpuIdDump {
+    type Item = (u32, Option<u32>, CpuIdResult);
+    type IntoIter = CpuIdDumpIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        CpuIdDumpIter {
+            dump: self,
+            leaf: 0,
+            current_subleaf: None,
+        }
+    }
+}
+
+impl Iterator for CpuIdDumpIter {
+    type Item = (u32, Option<u32>, CpuIdResult);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if let Some(subleaves) = self.current_subleaf.as_mut() {
+                if let Some(subleaf) = subleaves.keys().next().cloned() {
+                    let regs = subleaves.remove(&subleaf).expect("subleaf is present");
+                    return Some((self.leaf, Some(subleaf), regs));
+                } else {
+                    // We've exhauted this subleaf, move on.
+                    self.current_subleaf = None;
+                }
+            }
+
+            let Some(first_key) = self.dump.leaves.keys().next() else {
+                // We've exhausted the whole map!
+                return None;
+            };
+
+            self.leaf = *first_key;
+            let entry = self
+                .dump
+                .leaves
+                .remove(&self.leaf)
+                .expect("leaf is present");
+            match entry {
+                LeafOrSubleaves::Leaf(regs) => {
+                    return Some((self.leaf, None, regs));
+                }
+                LeafOrSubleaves::Subleaf(subleaves) => {
+                    self.current_subleaf = Some(subleaves);
+                }
+            }
+        }
+    }
+}
+
+const DEFAULT_LEAF: CpuIdResult = CpuIdResult {
+    eax: 0,
+    ebx: 0,
+    ecx: 0,
+    edx: 0,
+};
+
+impl CpuIdWriter for CpuIdDump {
+    fn set_leaf(&mut self, leaf: u32, mut bits: Option<CpuIdResult>) {
```

Comment:
`CpuIdWriter` requiring implementers to maintain internal consistency of the CPUID profile is not great, but I can imagine it being useful for the library to be able to represent *any* bit patterns in a CPUID profile. What's your preference on internally-inconsistent CPUID profiles? 

Either way, the "maintain internal consistency" logic here could be pulled out into a function in the library to support implementations, so external implementations are more like
```
fn set_leaf(...) {
    // update storage so that `.cpuid1()` and `.cpuid2()` queries reflect the new leaf
    self.update_state(leaf, bits);
    // update any other leaves as necessary with knowledge that `leaf` was set to `bits`?
    self.write_hook(leaf, bits);
}
```